### PR TITLE
Custom route for bulk api

### PR DESCRIPTION
**Pull request by nuzil**
Customisable rest route for Async Bulk API so that generated path can be overwritten 
Added XML file which declares the bulk API endpoint.

### CODE REVIEW

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/bulk-api-ce?pullRequest=21) <br/>All committers have signed the CLA.

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/bulk-api-ce?pullRequest=21) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dev.magento.com/magento/bulk-api-ce?pullRequest=21) before we can accept your contribution.<br/>

**Review by vrann — CHANGES_REQUESTED** [review]
There is already file webapi_async.xml, it would be better to reuse it, instead of adding a new one.
Another option is to add some attribute, i,e, "bulk" to the webapi.xml

@apaliarush what do you think?

File: app/code/Magento/WebapiAsync/Model/RouteCustomizationConfig.php [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+    /**
+     * Return route customizations loaded from cache if enabled or from files merged previously
+     *
+     * @return array
+     */
+    public function getRouteCustomizations()
+    {
+        if (null === $this->routeCustomizations) {
```
Comment (vrann) [inline app/code/Magento/WebapiAsync/Model/RouteCustomizationConfig.php]: config files are cached automatically, after structure is converted to array in Converter, there is no need to cache it again.

Comment (paliarush) [reply app/code/Magento/WebapiAsync/Model/RouteCustomizationConfig.php]: @vrann agree, we should re-use existing config. Also would be nice to allow override of this config parameter via deployment config (etc/env.php). See how it is done in message queue via custom config reader for deployment config.

**Comment by paliarush** [comment]
Does it make sense to introduce such feature for any web API routes?
If so, then it could be implemented in Webapi module directly without plugins.

**Comment by misha-kotov** [comment]
Yes, I think this functionality would provide value for all web API routes, not just Async/Bulk.

**Comment by nuzil** [comment]
@Neos2007 can you please check tests + implement this feature for ALL routes as described before

**Comment by Neos2007** [comment]
So custom routes logic must be implemented in Webapi module?
If that, for custom routes configuration a propose to add "custom_route" along with "route" nodes.
Will looks like: 
`<custom_route requested_path="asyncCustomers" endpoint="async/V1/customers"/> 
<custom_route requested_path="asyncBulkCmsPages" endpoint="async/bulk/V1/cmsPage" />`

**Comment by vrann** [comment]
@Neos2007 we discussed the concerns with @paliarush, so couple comments:
- we don't need to be able to rewrite it in env.php
- we don't need separate file to define overrides. Instead we propose to use `webapi_async.xml`
- the format of the new node:
```
<route url="" method="" alias="" />
```
All params should be required. Param `method` should allow only POST PUT DELETE GET
Nodes should be merged based on combination of `url` and `method` (composite identifier)

**Comment by Neos2007** [comment]
Modified logic of adding custom routes for asynchronous rest api.
Needed configuration for custom routes can be added to webapi_async.xml file as nodes inside root node.
Example:
`<route url="async/V1/products" method="POST" alias="asyncProducts" />`

File: app/code/Magento/WebapiAsync/Model/ServiceConfig/Converter.php [hunk 2]
```diff
@@ -20,8 +20,16 @@ class Converter implements \Magento\Framework\Config\ConverterInterface
     const KEY_METHOD = 'method';
     const KEY_METHODS = 'methods';
     const KEY_SYNCHRONOUS_INVOCATION_ONLY = 'synchronousInvocationOnly';
+    const KEY_ROUTES = 'routes';
     /**#@-*/
 
+    private $allowedRouteMethods = [
+        \Magento\Webapi\Model\Rest\Config::HTTP_METHOD_GET,
```
Comment (paliarush) [inline app/code/Magento/WebapiAsync/Model/ServiceConfig/Converter.php:27]: We should also allow `PATCH` for partial updates. It is stated in our web API technical vision https://devdocs.magento.com/guides/v2.2/coding-standards/technical-vision/webapi.html.

 Please add `PATCH` support for synchronous web APIs as well if not implemented yet and if time permits.

File: app/code/Magento/WebapiAsync/Model/ServiceConfig/Converter.php [hunk 3]
```diff
@@ -158,4 +167,69 @@ private function isSynchronousInvocationOnlyTrue(\DOMElement $synchronousInvocat
 
         return filter_var($synchronousInvocationOnlyNode->nodeValue, FILTER_VALIDATE_BOOLEAN);
     }
+
+    /**
+     * Convert and merge "route" nodes, which represent route customizations
```
Comment (paliarush) [inline app/code/Magento/WebapiAsync/Model/ServiceConfig/Converter.php:173]: Please correct formatting. 
Also, all new methods and classes must be documented across the review.

File: app/code/Magento/WebapiAsync/Plugin/ControllerRest.php [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    public function __construct(
+        ServiceConfig $serviceConfig,
+        PathProcessor $pathProcessor
+    ) {
+        $this->serviceConfig = $serviceConfig;
+        $this->pathProcessor = $pathProcessor;
+    }
+
+    /**
+     * Check is current rest api route path in route customization config.
```
Comment (paliarush) [inline app/code/Magento/WebapiAsync/Plugin/ControllerRest.php]: The description does not sound right. Please use something like "Apply route customization.", or get current version reviewed with English native speaker.

**Review by paliarush — CHANGES_REQUESTED** [review]
1. Would be nice to declare some real aliases for existing endpoints, these would serve as good examples (in addition to tests)
2. We should add at least one web API functional test for the implemented functionality

**Comment by nuzil** [comment]
If I understand correctly Tests are failed not because of PR

**Review by vrann — APPROVED** [review]

**Pull request merged**